stackit/internal/services/iaas/securitygrouprule/resource.go

```diff
@@ -0,0 +1,697 @@
+	"code": basetypes.Int64Type{},
+	"type": basetypes.Int64Type{},
+}
+
+type portRangeModel struct {
+	Max types.Int64 `tfsdk:"max"`
+	Min types.Int64 `tfsdk:"min"`
+}
+
+// Types corresponding to portRange
+var portRangeTypes = map[string]attr.Type{
+	"max": basetypes.Int64Type{},
+	"min": basetypes.Int64Type{},
+}
+
+type protocolModel struct {
+	Name     types.String `tfsdk:"name"`
+	Protocol types.Int64  `tfsdk:"protocol"`
+}
+
+// Types corresponding to protocol
+var protocolTypes = map[string]attr.Type{
+	"name":     basetypes.StringType{},
+	"protocol": basetypes.Int64Type{},
+}
+
+// NewSecurityGroupRuleResource is a helper function to simplify the provider implementation.
+func NewSecurityGroupRuleResource() resource.Resource {
+	return &securityGroupRuleResource{}
+}
+
+// securityGroupRuleResource is the resource implementation.
+type securityGroupRuleResource struct {
+	client *iaasalpha.APIClient
+}
+
+// Metadata returns the resource type name.
+func (r *securityGroupRuleResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_security_group_rule"
+}
+
+// Configure adds the provider configured client to the resource.
+func (r *securityGroupRuleResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	providerData, ok := req.ProviderData.(core.ProviderData)
+	if !ok {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error configuring API client", fmt.Sprintf("Expected configure type stackit.ProviderData, got %T", req.ProviderData))
+		return
+	}
+
+	if !resourceBetaCheckDone {
+		features.CheckBetaResourcesEnabled(ctx, &providerData, &resp.Diagnostics, "stackit_security_group_rule", "resource")
+		if resp.Diagnostics.HasError() {
+			return
+		}
+		resourceBetaCheckDone = true
+	}
+
+	var apiClient *iaasalpha.APIClient
+	var err error
+	if providerData.IaaSCustomEndpoint != "" {
+		ctx = tflog.SetField(ctx, "iaas_custom_endpoint", providerData.IaaSCustomEndpoint)
+		apiClient, err = iaasalpha.NewAPIClient(
+			config.WithCustomAuth(providerData.RoundTripper),
+			config.WithEndpoint(providerData.IaaSCustomEndpoint),
+		)
+	} else {
+		apiClient, err = iaasalpha.NewAPIClient(
+			config.WithCustomAuth(providerData.RoundTripper),
+			config.WithRegion(providerData.Region),
+		)
+	}
+
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error configuring API client", fmt.Sprintf("Configuring client: %v. This is an error related to the provider configuration, not to the resource configuration", err))
+		return
+	}
+
+	r.client = apiClient
+	tflog.Info(ctx, "iaasalpha client configured")
+}
+
+// Schema defines the schema for the resource.
+func (r *securityGroupRuleResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
```

Comment:
in general, can you check the openstack provider resource and see if it makes sense to have all these object fields instead of a more flat schema? https://registry.terraform.io/providers/terraform-provider-openstack/openstack/latest/docs/resources/networking_secgroup_rule_v2

You could also check this with the IaaS team.

Maybe something we could have checked in the refinement but we are still in time to make the best user experience here

Reply:
It would also remove a lot of the object mapping complexity, including that empty object case

Reply:
After aligning with the IaaS colleagues, we decided to leave them as objects

Reply:
Beware that some of these fields should only be set if a specific protocol is set. E.g:
- `icmp_parameters` **should only** be provided if the protocol is `icmp`
- `port_range` **shouldn't** be provided if protocol is `icmp`

We can't easily validate this in the schema I think, because it would require checking other fields values, but at least we should add to the description

Reply:
Added infos to the description. 
I used the examples you gave. Did not find any extra infos in API.